### PR TITLE
feat(pipelines transform): automatic namespacing

### PR DESCRIPTION
**Pull request by jdrouet**
## What is expected

A `pipelines` transform configuration can become a bit huge when the number of pipelines and transforms increases, making it harder to collaborate on it.
With the [automatic namespacing](https://github.com/vectordotdev/vector/pull/9571) feature, we can already move the transform configuration to a dedicated file but splitting it some more will improve the collaboration even more.

## Possible solutions

Let's consider the following config tree

```
+ /etc/vector
| + transforms
| | + processing
| | | + logs
| | | | + bar
| | | | | + first-transform.toml
| | | | | + second-transform.toml
| | | | + bar.json
| | | | + foo.toml
| | | + metrics
| | | | + baz.yaml
| | + processing.toml
| + vector.toml
```

We have to assume that in any case, considering that the code, when loading, has to know the fields and sub-directories it has to load or merge, we'll have to modify the `TransformConfig` from `vector_core` to add a method to do so.

Here are the considered solutions.

### Loading each transform and merging it into a `serde` value in order to deserialize it in a second time.

This implies:
  - Adding a `load_inner(&mut self, value: serde_toml::Value)`.
  - Loading the whole `processing` sub-directory when loading the `processing` transform and therefore not control the way we load the sub-directory. In that case, the `bar` sub-directory would load its content as a `Map` but, in the `PipelineConfig` structure, it's an array, so we would have to convert it manually.
  - Converting the different format into a common one (`json`, `toml` and `yaml` to `toml` for example) and therefore loose the possibility of specifying the file and line at which there probably is a validation error.

### Letting the responsibility to the transform to load its sub-folders from the filesystem.

This implies:
  - adding a `load_from_subdir(&mut self, path: &Path)` method to the `TransformConfig` trait with a default behavior for transform that don't support this splitting capability.
  - having a default value to most of the fields in order, for example, to load `bar.json` without its transforms and then load the transforms one by one.

## Final solution

The second one.

## TODO

- [x] Implement basic loading
- [x] Check that the pipelines can be in separate files and that they are treated alphabetically if no order is given
- [x] Check that the transforms can be in separate files and that they are loaded alphabetically

Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

**Review by lukesteensen — COMMENTED** [review]
I don't love the idea of coupling `TransformConfig` to the on-disk layout of config files, but I don't have a good alternative suggestion. I think it's reasonable to move forward with this for now and take what we learn towards improving our overall design when we take on the larger config rework task.

It looks like there's still a couple of compile issues in test, but otherwise this looks good to me.

**Comment by jszwedko** [comment]
Thanks @jdrouet ! Would you mind elaborating a bit on why the first solution doesn't work? It does seem like a natural extension, to me, to just extend the automatic namespacing feature we added before to work the same way for pipelines rather than having this transform need to know how to load configuration from files.

You mentioned:

> Loading the whole processing sub-directory when loading the processing transform and therefore not control the way we load the sub-directory. In that case, the bar sub-directory would load its content as a Map but, in the PipelineConfig structure, it's an array, so we would have to convert it manually.

I think it would be reasonable to not extend the automatic namespacing to configuration that is an array. That is, don't allow the `bar` pipeline to have its own nested directories. Otherwise, isn't it also unclear what order the transforms should be loaded in when loading from a subdirectory?

**Comment by jdrouet** [comment]
> Would you mind elaborating a bit on why the first solution doesn't work?

The main problem I see is that we'd loose the ability to point where the issues are in the configuration file (`error at line 8 of /etc/vector/transforms/processing/logs/...`) which is, to me, an important feature. And this is due to the fact the when loading all the files at the `ConfigBuilder` level we wouldn't know what kind of structure we are loading and we'd probably have to convert from on format to another if the configuration is mixed (`json` => `toml`).
Then, instead of having `/etc/vector/transforms/processing/logs/pipeline-a.toml` we'd have to have `/etc/vector/transforms/processing/logs/pipelines/pipeline-a.toml` to fit the configuration.

> I think it would be reasonable to not extend the automatic namespacing to configuration that is an array. That is, don't allow the bar pipeline to have its own nested directories. Otherwise, isn't it also unclear what order the transforms should be loaded in when loading from a subdirectory?

From what I remember from our meeting (I might be wrong) we ended up saying that we could load alphabetically the pipelines and the transforms. It seems like a better solution for the user rather than having everything in a single file.
If I remember well, one of the down sides of the `compound` transform was the lack of splitting in independent file. Then if we don't allow splitting the pipeline, we end up with the same down side.

**Comment by jszwedko** [comment]
> Then, instead of having `/etc/vector/transforms/processing/logs/pipeline-a.toml` we'd have to have `/etc/vector/transforms/processing/logs/pipelines/pipeline-a.toml` to fit the configuration.

I actually think the latter is preferable so that it directly maps to the configuration schema. This seems more intuitive to me as a user familiar with the "automatic namespacing" feature.

> From what I remember from our meeting (I might be wrong) we ended up saying that we could load alphabetically the pipelines and the transforms. It seems like a better solution for the user rather than having everything in a single file.
> If I remember well, one of the down sides of the `compound` transform was the lack of splitting in independent file. Then if we don't allow splitting the pipeline, we end up with the same down side.

We'd discussed alphabetically loading pipelines (and allowing the user to override with `order`) but I don't think we discussed further splitting out transforms in those pipelines into their own files. If we do do that, we should probably have the same `order` field on `PipelineConfig`. At this stage feature development for pipelines, I'm inclined to _not_ let users split out the pipeline transforms into their own files and just keep it at one file per pipeline. We can definitely revisit that in the future; I just think that is a simpler initial implementation.

**Comment by binarylogic** [comment]
> I'm inclined to not let users split out the pipeline transforms into their own files and just keep it at one file per pipeline

Agree. Individual files per transform is overkill and, as discussed here, it's not clear how transform ordering would work. Part of the appeal with the `pipelines` transform was to simplify the processing model. We don't want users thinking about connecting transforms or getting confused with ordering. If it's all in one file, it's simple and clear how the ordering works.

File: tests/namespacing/success/vector.toml [hunk 1]
```diff
@@ -1,6 +1,11 @@
 # Set global options
 data_dir = "/var/lib/vector"
```
Comment (vladimir-dd) [inline tests/namespacing/success/vector.toml:3]: can we add "processing" pipelines here as an input to any sink/transform? 
If I understand correctly now it's not connected to any output.

**Review by vladimir-dd — APPROVED** [review]
Looks good to me.

**Comment by jdrouet** [comment]
This implementation is replaced by https://github.com/vectordotdev/vector/pull/10179 to move the loading logic out of the components.

**Pull request closed without merging**